### PR TITLE
Yaml_parser document

### PR DESCRIPTION
**Pull request by kulrk**
# Problem Statement
-  Problem statement

# Design
-  For Bug, Describe the fix here.
-  For new test addition post the link for design

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is formatted (e.g. using pycharm in-built formatter)

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed
-  [ ] Attach test execution logs

# Review Checklist 
  Checklist for Author
-  [ ] JIRA number/GitHub Issue added to PR
-  [x] PR is self reviewed
-  [x] Jira and state/status is updated and JIRA is updated with PR link
-  [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
-  [x] Changes done to ReadMe

-----
[View rendered docs/yaml_structure.md](https://github.com/kulrk/corio/blob/yaml_parser_doc/docs/yaml_structure.md)

### CODE REVIEW

**Comment by VikasKumarC** [comment]
@kulrk  Also add one sample yaml file for reference to new users such that they can create workloads referring this doc and sample file.

File: docs/YAML_documents/sample_file.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+test_1:
+  TEST_ID: unique_string
```
Comment (ketanarlulkar) [inline docs/YAML_documents/sample_file.yaml]:
```suggestion
  TEST_ID: unique_string1
```

File: docs/YAML_documents/sample_file.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  object_size:
+    start: 0Kib
+    end: 100Kib
+  result_duration: 0d0h0m0s
+  sessions_per_node: 1
+  tool: s3api
+  operation: bucket
+
+test_2:
+  TEST_ID: unique_string
```
Comment (ketanarlulkar) [inline docs/YAML_documents/sample_file.yaml]:
```suggestion
  TEST_ID: unique_string2
```

File: docs/YAML_documents/sample_file.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  object_size:
+    start: 0Kb
+    end: 100Kb
+  result_duration: 0d0h0m0s
+  sessions_per_node: 2
+  tool: s3api
```
Comment (ketanarlulkar) [inline docs/YAML_documents/sample_file.yaml]: 1. Add example for part range start & end. Check `config/s3/s3api/multipart_random_object_size.yaml`
2. Add example for range_read. Check `config/s3/s3api/object_range_read.yaml`

File: docs/YAML_documents/yaml_structure.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+* test_id
+* object_size
+* part_size
+* result_duration
+* sessions_per_node
+* sessions
+* tool
+
+---
+
```
Comment (ketanarlulkar) [inline docs/YAML_documents/yaml_structure.md:16]: Add explanation for range_read key

**Comment by VikasKumarC** [comment]
@pranav-sahasrabudhe  @sonalk0209  Please review yaml sample file and related document.

File: docs/YAML_documents/yaml_structure.md (hunk 4 above)
Comment (VikasKumarC) [reply docs/YAML_documents/yaml_structure.md:16]: Please add explanation for every keyword. i.e. session_per_node

**Comment by VikasKumarC** [comment]
Hi @kulrk Replace/refactor result_duration with min_runtime in configs and code wherever needed

File: docs/YAML_documents/sample_file.yaml (hunk 3 above)
Comment (kulrk) [reply docs/YAML_documents/sample_file.yaml]: added

File: docs/YAML_documents/yaml_structure.md (hunk 4 above)
Comment (kulrk) [reply docs/YAML_documents/yaml_structure.md:16]: added

File: docs/YAML_documents/sample_file.yaml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  object_size:
+      start: 1Gib
+      end: 4Gib
+  part_range:
+      start: 50
+      end: 100
+  min_runtime: 2h
+  sessions: 5
+  tool: s3api
+  operation: multipart
```
Comment (ketanarlulkar) [inline docs/YAML_documents/sample_file.yaml]: sessions 5 and operation multipart how is it different than sessions_per_node 5 and operation multipart?

File: docs/YAML_documents/yaml_structure.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+TEST_ID can be any unique string which can be treated as filtering through execution logs.  
+
+object_size & part_size can be given as range for test workload, so those can have the following 
+keys 
+in it.
+* start
+* end
```
Comment (ketanarlulkar) [inline docs/YAML_documents/yaml_structure.md]:
```suggestion
object_size & part_size can be given as range for test workload, so those can have the following keys in it:
* start
* end
```

File: docs/YAML_documents/yaml_structure.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+* min_runtime
+* sessions_per_node
+* sessions
+* tool
+
+---
+
+TEST_ID can be any unique string which can be treated as filtering through execution logs.  
+
+object_size & part_size can be given as range for test workload, so those can have the following
```
Comment (ketanarlulkar) [inline docs/YAML_documents/yaml_structure.md]: Also object_size & part_size can be fixed size value. e.g. in config/s3/s3api/multipart_fixed_object_size.yaml
Also object_size & part_size can be a list of values. e.g. in config/s3/s3api/copy_object_fix_size.yaml

File: docs/YAML_documents/yaml_structure.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+Sessions_per_node and sessions is a multiplying factor which is used to calculate number of 
+sessions to be created within a test. Default value is 1.
+
+range_read key is used to perform range read operations in a workload. 
+
+part_range is range of numbers from which number of parts for multipart workloads can be 
+calculated. 
+
+Start and End range parameters are object/part sizes for which test will be executed.
+Sizes can be given from bytes, KB, MB up to TB. We can also use KiB format as well.
```
Comment (ketanarlulkar) [inline docs/YAML_documents/yaml_structure.md]:
```suggestion
Sizes can be given from bytes, KB (1000 Bytes), MB (1000 KB) up to TB (1000 GB). We can also use KiB (1024 Bytes) format as well. 
```

**Review by pranav-sahasrabudhe — APPROVED** [review]
Nothing much to review. Looks OK.

File: config/s3/s3api/bucket_operations.yaml [hunk 9]
```diff
@@ -3,7 +3,7 @@ test_1:
   object_size:
     start: 0Kib
     end: 100Kib
-  result_duration: 1h
+  min_runtime: 1h
```
Comment (sonalk0209) [inline config/s3/s3api/bucket_operations.yaml:6]: is there any max_runtime as well?

File: docs/YAML_documents/sample_file.yaml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  tool: s3api
+  operation: bucket
+
+test_2:
+  TEST_ID: unique_string2
+  object_size:
+    start: 0Kb
+    end: 100Kb
+  min_runtime: 0d0h0m0s
+  sessions_per_node: 2
```
Comment (sonalk0209) [inline docs/YAML_documents/sample_file.yaml:17]: what is the diff between sessions and sessions_per_node?

File: docs/YAML_documents/yaml_structure.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+* range_read
+* part_range
+* min_runtime
+* sessions_per_node
+* sessions
+* tool
+
+---
+
+TEST_ID can be any unique string which can be treated as filtering through execution logs.
```
Comment (sonalk0209) [inline docs/YAML_documents/yaml_structure.md:17]: its mentioned as lower case test_id above and here upper case, so which is to be used?

File: docs/YAML_documents/yaml_structure.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+range_read key is used to perform range read operations in a workload. 
+
+part_range is range of numbers from which number of parts for multipart workloads can be 
+calculated. 
+
+Start and End range parameters are object/part sizes for which test will be executed.
+Sizes can be given from bytes, KB, MB up to TB. We can also use KiB format as well. 
+
+Result duration can be specified from seconds up to days. For example: 1d1h, 1h or 2d1h2s.
```
Comment (sonalk0209) [inline docs/YAML_documents/yaml_structure.md]: isn't this min_runtime now?

File: docs/yaml_structure.md [hunk 13]
```diff
@@ -1,2 +0,0 @@
-
```
Comment (sonalk0209) [inline docs/yaml_structure.md:1]: what this file intended for? if nothing, please remove

File: config/s3/s3api/bucket_operations.yaml (hunk 9 above)
Comment (VikasKumarC) [reply config/s3/s3api/bucket_operations.yaml:6]: No

**Comment by VikasKumarC** [comment]
Hi @kulrk Check comment and resolve/reply

File: docs/yaml_structure.md (hunk 13 above)
Comment (kulrk) [reply docs/yaml_structure.md:1]: It is removed

File: docs/YAML_documents/yaml_structure.md (hunk 12 above)
Comment (kulrk) [reply docs/YAML_documents/yaml_structure.md]: fixed

File: docs/YAML_documents/yaml_structure.md (hunk 11 above)
Comment (kulrk) [reply docs/YAML_documents/yaml_structure.md:17]: Changed to upper case

File: docs/YAML_documents/sample_file.yaml (hunk 10 above)
Comment (kulrk) [reply docs/YAML_documents/sample_file.yaml:17]: Added some explanation in YAML sample file and in document, please have a look

**Review by ketanarlulkar — APPROVED** [review]

File: config/s3/s3api/bucket_operations.yaml (hunk 9 above)
Comment (sonalk0209) [reply config/s3/s3api/bucket_operations.yaml:6]: Then why not only runtime, this gives an impression that there is min and max runtime. just a suggestion.

File: docs/YAML_documents/yaml_structure.md [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+* start
+* end
+
+part_size can be given as fixed values and range of values for multipart workloads, so  those 
+can have the following keys in it.
+* start
+* end 
+
+Sessions_per_node and sessions is a multiplying factor which is used to calculate number of 
+sessions to be created within a test. Default value is 1. number of nodes param will be used to
```
Comment (sonalk0209) [inline docs/YAML_documents/yaml_structure.md]: number of nodes param is not listed

File: docs/YAML_documents/yaml_structure.md [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+* end
+
+part_size can be given as fixed values and range of values for multipart workloads, so  those 
+can have the following keys in it.
+* start
+* end 
+
+Sessions_per_node and sessions is a multiplying factor which is used to calculate number of 
+sessions to be created within a test. Default value is 1. number of nodes param will be used to 
+multiply with sessions per node and that total value will be used to create sessions and in case
```
Comment (sonalk0209) [inline docs/YAML_documents/yaml_structure.md]: *sessions_per_node

File: docs/YAML_documents/yaml_structure.md (hunk 14 above)
Comment (sonalk0209) [inline docs/YAML_documents/yaml_structure.md]: can we have explaination of sessions and sessions_per_node on 2 diff lines to avoid confusion

Comment (kulrk) [reply docs/YAML_documents/yaml_structure.md]: it is not a part of YAML file rather it is a command line argument, will try to specify more correctly.

File: config/s3/s3api/bucket_operations.yaml (hunk 9 above)
Comment (kulrk) [reply config/s3/s3api/bucket_operations.yaml:6]: this min_runtime specifies than minimum runtime for marking it pass. after specified time, next test in line will be called and existing running test case will continue to run.

File: docs/YAML_documents/yaml_structure.md [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+* range_read
+* part_range
+* min_runtime
+* sessions_per_node
+* sessions
+* tool
+
+---
+
+TEST_ID can be any unique string which can be treated as filtering through execution logs.
```
Comment (sonalk0209) [inline docs/YAML_documents/yaml_structure.md:17]: Maybe we can make all params bold as to ensure we are explaining same.

**Review by sonalk0209 — APPROVED** [review]
@kulrk just a minor comment.

**Comment by VikasKumarC** [comment]
@kulrk 
Resolve DCO check issue.

**Comment by VikasKumarC** [comment]
@kulrk  
Mark Conversations as resolved.

**Review by DhobaleVishal — APPROVED** [review]

**Pull request merged**